### PR TITLE
Dynamically render webmanifest file

### PR DESCRIPTION
**Pull request by ramsays**
The goal of this PR is to define a server rendered endpoint for the web app manifest file. The PR currently includes:

- Server-side rendering of `manifest.webmanifest` (icons are included in client PR)
  - Includes small icon for desktop browsers (Edge, Chrome)
  - Includes webp icons for Android (regular and maskable)
- Redirect for `favicon.ico` so legacy browsers have an icon (not referenced in `index.html`)

Left to do:

- [x] Determine how to render web manifest (runtime/build time)
- [x] Some browsers hit the root directory for `favicon.ico` and `apple-touch-icon`, how to manage?
- [x] Customize manifest theme color/icons for each school
- [x] Determine if we want to use PWA shortcuts (probably not, school dependent)

Preliminary review and cold-eyes appreciated. Testing should be done with `manifest` branch in client.

### CODE REVIEW

File: lib/frontend/views/manifest.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+		{
+			"src": "my/assets/android-chrome-512.png",
+			"sizes": "512x512",
+			"type": "image/png"
+		}
+	]
+}`;
+
+module.exports = function renderManifest(color) {
+	return template.replace(/__THEME_COLOR__/g, color);
```
Comment (vikaspotluri123) [inline lib/frontend/views/manifest.js:61]: Since this is a more popular view, I'd prefer to use string interpolation over replacements (`${var}` vs `string.replace(/__VAR__/g, var)`)

File: lib/services/school-config.js [hunk 2]
```diff
@@ -80,6 +79,15 @@ class SchoolConfigService {
 		return payload;
 	}
 
+	/**
+	 * @param {string} hostname the host of the requested theme
+	 * @returns {SchoolConfig} School configuration options
+	 */
+	getRawSchoolConfig(hostname) {
```
Comment (vikaspotluri123) [inline lib/services/school-config.js:86]: This should be an internal method, and we should expose a method for getting the primary color

File: lib/web/app.js [hunk 3]
```diff
@@ -57,6 +58,10 @@ module.exports.mount = app => {
 	app.use(hostMatching);
 	app.use(authentication.withUser);
 
+	app.get('/manifest.webmanifest', pwa.getManifest);
```
Comment (vikaspotluri123) [inline lib/web/app.js:61]:
```suggestion
	app.get('/api/v0/web-manifest', pwa.getManifest);
```

You might also have to set `content-type` headers

**Review by vikaspotluri123 — CHANGES_REQUESTED** [review]
Good start 😄 

As an aside, the client submodule should only be updated for releases. The submodule tag is used to track what client release is guaranteed to work with a given server release

File: lib/controllers/index.js [hunk 4]
```diff
@@ -7,6 +7,7 @@ module.exports = {
 	home: require('./home'),
 	middleware: require('./middleware'),
 	semester: require('./semester'),
+	pwa: require('./pwa'),
```
[the remark this answers is not in the document]
Comment (ramsays) [reply lib/controllers/index.js]: No longer needed 😁

File: lib/services/school-config.js (hunk 2 above)
Comment (ramsays) [reply lib/services/school-config.js]: No longer needed 😁

File: lib/web/app.js (hunk 3 above)
Comment (ramsays) [reply lib/web/app.js]: No longer needed 😁

File: lib/frontend/views/manifest.js (hunk 1 above)
Comment (ramsays) [reply lib/frontend/views/manifest.js]: No longer needed! 👍

**Comment by ramsays** [comment]
Because school config is now doing all the heavy lifting, server scope went from interesting to basically trivial. LGTM!

**Comment by vikaspotluri123** [comment]
Do we want to move the dynamically computed theme colors here as well? Now that we actually have a decent chunk of markup being dynamically rendered?

**Comment by vikaspotluri123** [comment]
Just to make sure we have the same expectations for client, I'll review this once client is ready. For now it looks good

File: lib/utils/browser-injects.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+	<link rel="apple-touch-startup-image" href="https://static.gradebook.app/sYbR9JXKTI/apple-splash-ipad-lg-horiz.png" media="(device-width: 1112px) and (device-height: 834px) and (-webkit-device-pixel-ratio: 2)" />
+	<link rel="apple-touch-startup-image" href="https://static.gradebook.app/sYbR9JXKTI/apple-splash-ipad-xl.png" media="(device-width: 834px) and (device-height: 1194px) and (-webkit-device-pixel-ratio: 2)" />
+	<link rel="apple-touch-startup-image" href="https://static.gradebook.app/sYbR9JXKTI/apple-splash-ipad-xl-horiz.png" media="(device-width: 1194px) and (device-height: 834px) and (-webkit-device-pixel-ratio: 2)" />
+	<link rel="apple-touch-startup-image" href="https://static.gradebook.app/sYbR9JXKTI/apple-splash-ipad-2xl.png" media="(device-width: 1024px) and (device-height: 1366px) and (-webkit-device-pixel-ratio: 2)" />
+	<link rel="apple-touch-startup-image" href="https://static.gradebook.app/sYbR9JXKTI/apple-splash-ipad-2xl-horiz.png" media="(device-width: 1366px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2)" />`;
+	}
+
+	return '';
+};
```
Comment (vikaspotluri123) [inline lib/utils/browser-injects.js:36]: This should be renamed to `apple-meta-tax.js` (or `apple-meta.js`) and needs tests using UAs from iOS, iPadOS, Safari, Edge, ChrEdge, Chrome, Firefox, and Brave (along with any others you can think of)

```suggestion
// @ts-check
const CDN_ROOT = 'https://static.gradebook.app/sYbR9JXKTI/';
const APPLE_SPECIFIC_META_TAGS = `
<meta name="apple-touch-fullscreen" content="yes" />
<meta name="apple-mobile-web-app-title" content="Gradebook" />
<meta name="apple-mobile-web-app-capable" content="yes" />
<meta name="apple-mobile-web-app-status-bar-style" content="default" />
<link rel="apple-touch-startup-image" href="${CDN_ROOT}apple-splash-iphone.png" media="(device-width: 375px) and (device-height: 667px) and (-webkit-device-pixel-ratio: 2)" />
<link rel="apple-touch-startup-image" href="${CDN_ROOT}/apple-splash-iphone-plus.png" media="(device-width: 414px) and (device-height: 736px) and (-webkit-device-pixel-ratio: 3)" />
<link rel="apple-touch-startup-image" href="${CDN_ROOT}/apple-splash-iphone-x.png" media="(device-width: 375px) and (device-height: 812px) and (-webkit-device-pixel-ratio: 3)" />
<link rel="apple-touch-startup-image" href="${CDN_ROOT}/apple-splash-iphone-xr.png" media="(device-width: 414px) and (device-height: 896px) and (-webkit-device-pixel-ratio: 2)" />
<link rel="apple-touch-startup-image" href="${CDN_ROOT}/apple-splash-iphone-max.png" media="(device-width: 414px) and (device-height: 896px) and (-webkit-device-pixel-ratio: 3)" />
<link rel="apple-touch-startup-image" href="${CDN_ROOT}/apple-splash-iphone-12.png" media="(device-width: 390px) and (device-height: 844px) and (-webkit-device-pixel-ratio: 3)" />
<link rel="apple-touch-startup-image" href="${CDN_ROOT}/apple-splash-iphone-12max.png" media="(device-width: 428px) and (device-height: 926px) and (-webkit-device-pixel-ratio: 3)" />
<link rel="apple-touch-startup-image" href="${CDN_ROOT}/apple-splash-ipad-sm.png" media="(device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2)" />
<link rel="apple-touch-startup-image" href="${CDN_ROOT}/apple-splash-ipad-sm-horiz.png" media="(device-width: 1024px) and (device-height: 768px) and (-webkit-device-pixel-ratio: 2)" />
<link rel="apple-touch-startup-image" href="${CDN_ROOT}/apple-splash-ipad-md.png" media="(device-width: 810px) and (device-height: 1080px) and (-webkit-device-pixel-ratio: 2)" />
<link rel="apple-touch-startup-image" href="${CDN_ROOT}/apple-splash-ipad-md-horiz.png" media="(device-width: 1080px) and (device-height: 810px) and (-webkit-device-pixel-ratio: 2)" />
<link rel="apple-touch-startup-image" href="${CDN_ROOT}/apple-splash-ipad-lg.png" media="(device-width: 834px) and (device-height: 1112px) and (-webkit-device-pixel-ratio: 2)" />
<link rel="apple-touch-startup-image" href="${CDN_ROOT}/apple-splash-ipad-lg-horiz.png" media="(device-width: 1112px) and (device-height: 834px) and (-webkit-device-pixel-ratio: 2)" />
<link rel="apple-touch-startup-image" href="${CDN_ROOT}/apple-splash-ipad-xl.png" media="(device-width: 834px) and (device-height: 1194px) and (-webkit-device-pixel-ratio: 2)" />
<link rel="apple-touch-startup-image" href="${CDN_ROOT}/apple-splash-ipad-xl-horiz.png" media="(device-width: 1194px) and (device-height: 834px) and (-webkit-device-pixel-ratio: 2)" />
<link rel="apple-touch-startup-image" href="${CDN_ROOT}/apple-splash-ipad-2xl.png" media="(device-width: 1024px) and (device-height: 1366px) and (-webkit-device-pixel-ratio: 2)" />
<link rel="apple-touch-startup-image" href="${CDN_ROOT}/apple-splash-ipad-2xl-horiz.png" media="(device-width: 1366px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2)" />
`;

/**
* @param {Gradebook.Request} request
*/
module.exports = request => {
	const ua = request.headers['user-agent'];
	const hasSafari = ua.includes('Safari');
	const hasChrome = ua.includes('Chrom');
	const hasEdge = ua.includes('Edg');

	return (hasSafari && !hasChrome && !hasEdge) ? APPLE_SPECIFIC_META_TAGS : '';
};

```

**Comment by ramsays** [comment]
A few comments on the user agent tests:

- Brave (and other smaller browsers) copy Chromes user agent string
- I didn't include MSEdge because it is being quickly superseded by the new Edge
- iPad and Mac now use the same user agent string (as of iOS 13+), so only one test is available, even though we would desire to not render the apple tags on Safari desktop

File: lib/utils/apple-meta.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+* @param {Gradebook.Request} request
+*/
+module.exports = request => {
+	const ua = request.headers['user-agent'];
+	const hasSafari = ua.includes('Safari');
+	const hasChrome = ua.includes('Chrom');
+	const hasEdge = ua.includes('Edg');
+
+	// Special injects for Safari
+	return (hasSafari && !hasChrome && !hasEdge) ? APPLE_SPECIFIC_META_TAGS : '';
```
Comment (joshcosta) [inline lib/utils/apple-meta.js:36]: Do we need to check for Chrome and Edge here?

Comment (vikaspotluri123) [reply lib/utils/apple-meta.js:36]: Based on the current algorithm, yes, because Chrome includes Safari in their UA.

My preferred algo is matching `/iphone|ipod|mac/i`

**Comment by joshcosta** [comment]
Before release we need to make sure the assets are up on static.

File: lib/utils/apple-meta.js (hunk 6 above)
Comment (ramsays) [reply lib/utils/apple-meta.js]: I checked that and yeah the regex should work.

File: test/unit/utils/apple-meta.spec.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+const safariPhoneUA = 'Mozilla/5.0 (iPhone; CPU iPhone OS 14_0_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Mobile/15E148 Safari/604.1';
+const safariPadUA = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.1 Safari/605.1.15';
+
+const requestStub = userAgent => {
+	return {headers: {'user-agent': userAgent}};
+};
+
+describe('Unit > Utils > Apple Meta', function () {
+	it('Safari on iOS shows custom meta', function () {
+		expect(appleMeta(requestStub(safariPhoneUA)).length).to.not.be.equal(0);
```
[the remark this answers is not in the document]
Comment (ramsays) [reply test/unit/utils/apple-meta.spec.js]: Indeed 👍

**Review by vikaspotluri123 — APPROVED** [review]
